src/abe/dippe.c

```diff
@@ -0,0 +1,455 @@
+    // check whether there is the same number of keys as attribute vector components
+    if (av->len != usks_len) {
+        return CFE_ERR_INSUFFICIENT_KEYS;
+    }
+
+    // temp vars
+    ECP_BN254 tmp_g1;
+    ECP2_BN254 tmp_g2;
+    FP12_BN254 tmp_gt;
+
+    // C0_K
+    FP12_BN254 C0_K;
+    FP12_BN254_one(&C0_K);
+    for (size_t i = 0; i < (dippe->assump_size + 1); i++) {
+        ECP2_BN254_inf(&tmp_g2);
+        for (size_t j = 0; j < usks_len; j++) {
+            ECP2_BN254_add(&tmp_g2, &((usks + j)->Ki.vec[i]));
+        }
+        PAIR_BN254_ate(&tmp_gt, &tmp_g2, &(cipher->C0.vec[i]));
+        PAIR_BN254_fexp(&tmp_gt);
+        FP12_BN254_mul(&C0_K, &tmp_gt);
+    }
+
+    // Ci_H
+    FP12_BN254 Ci_H;
+    FP12_BN254_one(&Ci_H);
+    char buf[cfe_dippe_int_len(dippe->assump_size + 1) + strlen(gid) + strlen(av->s) + 3];
+    for (size_t i = 0; i < (dippe->assump_size + 1); i++) {
+        ECP_BN254_inf(&tmp_g1);
+        for (size_t j = 0; j < usks_len; j++) {
+            if (mpz_tstbit(av->attrs, j) == 1) {
+                ECP_BN254_add(&tmp_g1, &(cipher->Ci.mat[j].vec[i]));
+            }
+        }
+        sprintf(buf, "%zu|%s|%s", i, gid, av->s);
+        cfe_dippe_hash_G2(&tmp_g2, buf, strlen(buf));
+        PAIR_BN254_ate(&tmp_gt, &tmp_g2, &tmp_g1);
+        PAIR_BN254_fexp(&tmp_gt);
+        FP12_BN254_mul(&Ci_H, &tmp_gt);
+    }
+
+    // C0_K * Ci_H
+    FP12_BN254_copy(&tmp_gt, &C0_K);
+    FP12_BN254_mul(&tmp_gt, &Ci_H);
+
+    // recover secret
+    FP12_BN254_inv(&tmp_gt, &tmp_gt);
+    FP12_BN254_copy(result, &(cipher->C_prime));
+    FP12_BN254_mul(result, &tmp_gt);
+
+    return CFE_ERR_NONE;
+}
+
+cfe_error cfe_dippe_attribute_vector_init(cfe_dippe_attribute_vector *av, size_t vec_len, size_t pattern[], size_t pat_len) {
+    mpz_init(av->attrs);
+    mpz_set_ui(av->attrs, 0);
+    for (size_t i = 0; i < pat_len; i++) {
+        if (pattern[i] < vec_len) {
+            mpz_setbit(av->attrs, pattern[i]);
+        } else
+            return CFE_ERR_MALFORMED_INPUT;
+    }
+    mpz_setbit(av->attrs, (vec_len - 1));
+
+    // attribute vector as string
+    av->s = (char *)cfe_malloc((mpz_sizeinbase(av->attrs, 62) + 1) * sizeof(char));
+    mpz_get_str(av->s, 62, av->attrs);
+    // attribute vector length
+    av->len = vec_len;
+
+    return CFE_ERR_NONE;
+}
+
+void cfe_dippe_attribute_vector_free(cfe_dippe_attribute_vector *av) {
+    mpz_clear(av->attrs);
+    free(av->s);
+}
+
+cfe_error cfe_dippe_threshold_policy_vector_init(cfe_dippe_policy_vector *pv, cfe_dippe *dippe, size_t vec_len, size_t pattern[], size_t pat_len, size_t threshold) {
+    mpz_t zero;
+    mpz_init(zero);
+    mpz_set_ui(zero, 0);
+    cfe_vec_init(&(pv->pol), vec_len);
+    cfe_vec_set_const(&(pv->pol), zero);
+
+    for (size_t i = 0; i < pat_len; i++) {
+        if (pattern[i] < vec_len) {
```

Comment:
I believe that it should be `< vec_len - 1`. Similarly in the conjunction function and in `cfe_dippe_attribute_vector_init`.